config/src/config/override_node_config.rs

```diff
@@ -0,0 +1,299 @@
+// Copyright © Aptos Foundation
+
+use crate::config::{Error, NodeConfig, PersistableConfig};
+use anyhow::bail;
+use serde::{Deserialize, Serialize};
+use std::path::Path;
+
+/// Diff a config yaml with a base config yaml. Returns None if there is no diff.
+fn diff_override_config_yaml(
```

Comment:
This function can be made a bit cleaner and reduce code duplication by writing as 

```
fn diff_override_config_yaml(
    override_config: serde_yaml::Value,
    base_config: serde_yaml::Value,
) -> anyhow::Result<Option<serde_yaml::Value>> {
    match (override_config, base_config) {
        (serde_yaml::Value::Mapping(override_mapping), serde_yaml::Value::Mapping(base_mapping)) => {
            let mut overrides = serde_yaml::Mapping::new();
            for (override_key, override_value) in override_mapping {
                if let Some(base_value) = base_mapping.get(&override_key) {
                    if let Some(diff_value) = diff_override_config_yaml(override_value, base_value.clone())? {
                        overrides.insert(override_key, diff_value);
                    }
                } else {
                    overrides.insert(override_key, override_value);
                }
            }
            if overrides.is_empty() {
                Ok(None)
            } else {
                Ok(Some(serde_yaml::Value::Mapping(overrides)))
            }
        },
        (serde_yaml::Value::Null, serde_yaml::Value::Null) => Ok(None),
        (serde_yaml::Value::Bool(override_value), serde_yaml::Value::Bool(base_value)) => {
            if override_value == base_value {
                Ok(None)
            } else {
                Ok(Some(override_config))
            }
        },
        (serde_yaml::Value::Number(override_value), serde_yaml::Value::Number(base_value)) => {
            if override_value == base_value {
                Ok(None)
            } else {
                Ok(Some(override_config))
            }
        },
        (serde_yaml::Value::String(override_value), serde_yaml::Value::String(base_value)) => {
            if override_value == base_value {
                Ok(None)
            } else {
                Ok(Some(override_config))
            }
        },
        (serde_yaml::Value::Sequence(override_value), serde_yaml::Value::Sequence(base_value)) => {
            if override_value == base_value {
                Ok(None)
            } else {
                Ok(Some(override_config))
            }
        },
        _ => bail!("base does not match override: {:?}, {:?}", override_config, base_config),
    }
}

```

Reply:
You can go more fancier and reduce the duplication further with use of macro as well..


```
macro_rules! compare_and_return {
    ($override_value:expr, $base_value:expr, $override_config:expr, $type_name:literal) => {{
        if $override_value == $base_value {
            Ok(None)
        } else {
            bail!(
                "base does not match override: {}({}), {:?}",
                $type_name,
                $override_value,
                $base_value
            );
        }
    }};
}

fn diff_override_config_yaml(
    override_config: serde_yaml::Value,
    base_config: serde_yaml::Value,
) -> anyhow::Result<Option<serde_yaml::Value>> {
    match (&override_config, &base_config) {
        (
            serde_yaml::Value::Mapping(override_mapping),
            serde_yaml::Value::Mapping(base_mapping),
        ) => {
            let mut overrides = serde_yaml::Mapping::new();
            for (override_key, override_value) in override_mapping {
                if let Some(base_value) = base_mapping.get(override_key) {
                    if let Some(diff_value) =
                        diff_override_config_yaml(override_value.clone(), base_value.clone())?
                    {
                        overrides.insert(override_key.clone(), diff_value);
                    }
                } else {
                    overrides.insert(override_key.clone(), override_value.clone());
                }
            }
            if overrides.is_empty() {
                Ok(None)
            } else {
                Ok(Some(serde_yaml::Value::Mapping(overrides)))
            }
        }
        (serde_yaml::Value::Null, serde_yaml::Value::Null) => Ok(None),
        (serde_yaml::Value::Bool(override_value), serde_yaml::Value::Bool(base_value)) => {
            compare_and_return!(override_value, base_value, override_config, "Bool")
        }
        (serde_yaml::Value::Number(override_value), serde_yaml::Value::Number(base_value)) => {
            compare_and_return!(override_value, base_value, override_config, "Number")
        }
        (serde_yaml::Value::String(override_value), serde_yaml::Value::String(base_value)) => {
            compare_and_return!(override_value, base_value, override_config, "String")
        }
        (serde_yaml::Value::Sequence(override_value), serde_yaml::Value::Sequence(base_value)) => {
            compare_and_return!(override_value, base_value, override_config, "Sequence")
        }
        _ => bail!(
            "base does not match override: {:?}, {:?}",
            override_config,
            base_config
        ),
    }
}

```

Reply:
I think that changes the behavior of when override is a Mapping but the base is not. But let me see if I can still simplify it.